pcdsdevices/pim.py

```diff
@@ -30,39 +28,197 @@ class PIMMotor(InOutRecordPositioner):
+
+    def remove(self, moved_cb=None, timeout=None, wait=False):
+        """Moves the diode out of the beam."""
+        return self.state.remove(moved_cb=moved_cb, timeout=timeout,
+                                 wait=wait)
+
+    def __init__(self, prefix, *, name, prefix_det=None, prefix_zoom=None,
+                 **kwargs):
+        self.infer_prefix(prefix)
+
+        # Infer the detector PV from the base prefix
+        if prefix_det:
             self._prefix_det = prefix_det
+        else:
+            self._prefix_det = self.prefix_start+'CVV:01'
+
+        # Infer the zoom motor PV from the base prefix
+        if prefix_zoom:
+            self._prefix_zoom = prefix_zoom
+        else:
+            self._prefix_zoom = self.prefix_start+'CLZ:01'
+
+        super().__init__(prefix, name=name, **kwargs)
+        self.y_motor = self.state.motor
+
+
+class PIM_withFocus(PIM):
+    """
+    Profile intensity monitor with y-motion motor, zoom motor, focus motor, and
+    a detector.
+
+    Parameters
+    ----------
+    prefix : str
+        The EPICS base of the PIM
+
+    name : str
+        A name to refer to the device
+
+    prefix_det : str, optional
+        The EPICS base PV of the detector. If None, it will be attemptted to be
+        inferred from `prefix`
+
+    prefix_zoom : str, optional
+        The EPICS base PV of the zoom motor. If None, it will be attempted to
+        be inferred from `prefix`
+
+    prefix_focus : str, optional
+        The EPICS base PV of the focus motor. If None, it will be attempted to
+        be inferred from `prefix`
+    """
+    focus_motor = FCpt(IMS, '{self._prefix_focus}')
+
+    def __init__(self, prefix, *, name, prefix_focus=None, **kwargs):
+        self.infer_prefix(prefix)
+
+        # Infer the focus motor PV from the base prefix
+        if prefix_focus:
+            self._prefix_focus = prefix_focus
+        else:
+            self._prefix_focus = self.prefix_start+'CLF:01'
+
         super().__init__(prefix, name=name, **kwargs)
+
+
+class PIM_withLED(PIM):
+    """
+    Profile intensity monitor with y-motion motor, zoom motor, LED, and a
+    detector.
+
+    Parameters
+    ----------
+    prefix : str
+        The EPICS base of the PIM
+
+    name : str
+        A name to refer to the device
+
+    prefix_det : str, optional
+        The EPICS base PV of the detector. If None, it will be attemptted to be
+        inferred from `prefix`
+
+    prefix_zoom : str, optional
+        The EPICS base PV of the zoom motor. If None, it will be attempted to
+        be inferred from `prefix`
+
+    prefix_led : str, optional
+        The EPICS base PV of the LED. If None, it will be attempted to be
+        inferred from `prefix`
+    """
+    led = FCpt(EpicsSignal, '{self._prefix_led}')
+
+    def __init__(self, prefix, *, name, prefix_led=None, **kwargs):
+        self.infer_prefix(prefix)
+
+        # Infer the illuminator PV from the base prefix
```

Comment:
Just an idea, but I wonder if all of the infer steps be in a customized `infer_prefix` method in the subclasses.

e.g.,
```python
def __init__(self, prefix, ..., **kwargs):
    # Stash away all `prefix_*` kwargs by convention
    self._prefixes = {key.split('prefix_')[1]: kwargs.pop(key) for key in kwargs if key.startswith('prefix_')}
    self.infer_prefix()
    super().__init__(...)

def infer_prefix(self):
    super().infer_prefix()
    # Infer the illuminator PV from the base prefix
    self._prefixes['led']
```

Maybe it's useful as it seems like we repeat this sort of code a lot, but maybe it's too confusing...

Reply:
I'd like to do this, but I can't figure out a simple way to make it work.

Reply:
Zach and I tried to figure it out to no avail and also noted that such a change would also make the code less readable.

Reply:
:+1: